### ISSUE TITLE
[question] How to model header only libraries that have a runtime dependency?

### ISSUE THREAD

**Issue by foundry-markf** [conan-io/conan#8314]
Hi,

Tha basic question is which model to use for recipes that are header only libraries that have a dependency at runtime on a library? It doesn't get covered in https://docs.conan.io/en/latest/howtos/header_only.html.

A example of this is cppzmq, as it's modelled here: https://github.com/conan-io/conan-center-index/blob/master/recipes/cppzmq/all/conanfile.py

It has `settings`, and `requires` (on ZeroMQ), but package_id wipes all that ABI information out with `self.info.header_only()`.

Looking at the Conan Center Artifactory (https://conan.io/center/cppzmq?tab=configuration&os=) there is a single build there (expected for header only).

However, in the more general sense, if we wanted to write a test_package recipe for a header only library that needs to exercise different variations of its dependent runtime library, how is that meant to be modelled? For example, say if H is the header only lib, that requires the runtime lib R, but we know we have different needs to use it in both shared=True and shared=False modes of R, how is that modelled in a recipe for H and make it testable at package creation time?

Obviously, it would be nice not to have duplicates of H for every variation of R we're interested in, but that does seem like the only solution, and not to use header_only().

Thanks!
Mark

- [x ] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by foundry-markf** [comment]
Does anyone have any thoughts on this? thanks

**Comment by fschoenm** [comment]
Did you ever find a solution for that? I think we have the same problem with asio and runtime dependencies that are OS-dependent (pthreads).

**Comment by foundry-markf** [comment]
@fschoenm No, I never heard anything back at all to clarify what we should be doing.

Although, now I've checked, the conan-center recipe for cppzmq has been subsequently updated to remove some of issues I raised in https://github.com/conan-io/conan-center-index/commit/f192097674b272ae20b0603d053adf14ad313acf#diff-d5da3b5b050ac6af64196a95226d6df0c6989fef9091d9d586e0f9bfcee583c6 . The runtime requirement remains, but self.info.header_only() is still used, so it's removed from the package_id.

Ideally, we want a single package_id, but test N times against different runtime package_ids.

**Comment by memsharded** [comment]
Hi!

> However, in the more general sense, if we wanted to write a test_package recipe for a header only library that needs to exercise different variations of its dependent runtime library, how is that meant to be modelled? For example, say if H is the header only lib, that requires the runtime lib R, but we know we have different needs to use it in both shared=True and shared=False modes of R, how is that modelled in a recipe for H and make it testable at package creation time?

It depends on how you want to transmit that dependency to consumers. It is correct to do something like:

```python
def package_info(self):
     if self.settings.os == "Windows":
           self.cpp_info.system_libs = ["winpthread"]     
     else:
           self.cpp_info.system_libs = ["pthread"]    

def package_id(self):
     self.info.header_only()
```

Even if the information is wiped from the ``package_id``, the information that consumers get can still be different based on settings and options.

Is this the behavior you are interested in?

**Comment by foundry-markf** [comment]
Hi @memsharded,

Thanks for the reply.

The model that we're ideally interested in is probably more vague than that. We have several products, all interested in using header only libs (with runtime requirements), but each product can use different flavours of the runtime requirements. One, for example, uses a static lib while another uses a shared lib.

So, while we don't want the header library to have ABI, i.e. don't say anything about particular runtime requirements (leaving that decision entirely to the consumer), we do want, at package creation time, to verify that it will work with all the variants we know about for the runtime requirement. Hence, the need for a test package to run several times against different variants.

Currently, I think that's on us, to write into CI, and run `conan test` several times.

So the question is more about is there a correct way to model this in recipes.

Mark

**Comment by memsharded** [comment]
> So, while we don't want the header library to have ABI, i.e. don't say anything about particular runtime requirements (leaving that decision entirely to the consumer), we do want, at package creation time, to verify that it will work with all the variants we know about for the runtime requirement. Hence, the need for a test package to run several times against different variants.

> Currently, I think that's on us, to write into CI, and run conan test several times.

Ok I think I understand a bit better. I guess the alternatives are:

- If the tests are "package tests" in the ``test_package``, then yes, the correct way would be to call ``conan test`` or ``conan create`` as many times as desired with the different input profiles and configurations. You might use the ``--build=missing`` in case you want to avoid actually re-building the header-only package itself.
- If the tests are inside the package recipe itself (in case you wanted to test some internals that wouldn't be as exposed from outside of the package), then it would be necessary to actually re-build the package for every configuration, even if it results in the same ``package_id``.

So what you said, running multiple ``conan test`` or ``conan create`` for the different configurations seems the way to go. There is nothing in recipes for this case, as when recipes evaluate they always process 1 configuration, so it is not possible to test more than one in a single pass

**Comment by foundry-markf** [comment]
OK, good to know we haven't missed anything obvious for markup and intent.

We'll approach it making header libs ABI-free, even when there are runtime requirements as a detail to the consumer, and work on our CI to do the right variations in testing.

**Comment by memsharded** [comment]
Closing this ticket as old and outdated (Conan 1.X), it is likely that Conan 2 new model with requirement traits, package types, etc, could model this case a bit better.

Please create new tickets for any further question or issue, thanks for the feedback!
